### DISCUSSION TITLE
CCTV switching day/night profile from tasmota

### DISCUSSION THREAD

**Discussion by rapi3** [arendst/Tasmota#14916] [Q&A] [answered]
Hello,
I want to use tasmota to switch day / night profile in Dahua camera.
All I need is to send this commands from browser to work - tested OK:
```
#day profile
http://user:pass@192.168.12.166/cgi-bin/configManager.cgi?action=setConfig&VideoInMode[0].Config[0]=0
#night profile
http://user:pass@192.168.12.166/cgi-bin/configManager.cgi?action=setConfig&VideoInMode[0].Config[0]=1
```
I tried with **websend** and **webquery** command from tasmota console but luck, no changes on camera and no error on console :(
Any ideas ?
Thank you.

**Comment by ascillato** [comment, marked as answer]
Can you share your exact rule?

Have you encoded the url?

**Reply by rapi3** [reply]
At this moment I don't have a rule I just test the webquery/websend commands from tasmota console to be sure it is working properly... and I can't find the right one.
this is log from linux terminal for command sent with wget:
```
wget 'http://admin:xxxxxxxx@192.168.12.166/cgi-bin/configManager.cgi?action=setConfig&VideoInMode[0].Config[0]=1'
--2022-02-19 23:15:45--  http://admin:*password*@192.168.12.166/cgi-bin/configManager.cgi?action=setConfig&VideoInMode[0].Config[0]=1
Connecting to 192.168.12.166:80... connected.
HTTP request sent, awaiting response... 401 Unauthorized
Syntax error in Set-Cookie: secure; HttpOnly at position 7.
Authentication selected: Digest realm="Login to b1fb517e19c782ac15a8ee79e124e542", qop="auth", nonce="23223391", opaque="21ba2cc2281212d9bc5762e3f10966b56bf2e164"
Connecting to 192.168.12.166:80... connected.
HTTP request sent, awaiting response... 200 OK
Syntax error in Set-Cookie: secure; HttpOnly at position 7.
Length: 4 [text/plain]
Saving to: ‘configManager.cgi?action=setConfig&VideoInMode[0].Config[0]=1’

configManager.cgi?action=setConfig 100%[================================================================>]       4  --.-KB/s    in 0s      

2022-02-19 23:15:45 (881 KB/s) - ‘configManager.cgi?action=setConfig&VideoInMode[0].Config[0]=1’ saved [4/4]
```

**Reply by sfromis** [reply]
Well, when you have issues with Tasmota, better post the log from Tasmota, which would include the exact command you're trying to make work. Anyway, the response of `401 Unauthorized` suggests that it does not work right away with wget either.... Looks like wget is doing something automatic about authentication, which is not part of what Tasmota can do.

**Reply by ascillato** [reply]
what about:

```
WebQuery http%3A%2F%2Fadmin%3Axxxxxxxx%40192.168.12.166%2Fcgi-bin%2FconfigManager.cgi%3Faction%3DsetConfig%26VideoInMode%5B0%5D.Config%5B0%5D%3D1 GET
```

That is the same as: 
```
WebQuery http://admin:xxxxxxxx@192.168.12.166/cgi-bin/configManager.cgi?action=setConfig&VideoInMode[0].Config[0]=1 GET
```
but URLencoded using https://www.urlencoder.org/

**Reply by rapi3** [reply]
@sfromis 
The problem is that I don't have any error log all commands are sent with result **{"WebQuery":"Done"}**
The problem can be the digest authentication required for login ??

**Reply by rapi3** [reply]
@ascillato
unfortunately it is not working with command encoded and I have this error **{"WebQuery":"Host not found"}**

**Reply by sfromis** [reply]
You still did not post what you did. And yes, I already pointed out that the authentication flow you see with wget will not be done by Tasmota. You need an URL working in **one** step, without any retries, redirects, cookies or whatever.

**Reply by rapi3** [reply]
ok, so no chance to do this from tasmota... :(
thank you
```
curl --digest -u "admin:xxxxxxxxx" -g "http://192.168.12.166/cgi-bin/configManager.cgi?action=setConfig&VideoInMode[0].Config[0]=1"
OK
```

**Reply by barbudor** [reply]
You're giving up to easily
If there is redirection, you should use 'curl -v' to find out what is the final url and work with it
Authentication should be provide as a header  'Authorization: Basic XXXX' where XXXX is the string 'user:password' encoded base64

**Reply by rapi3** [reply]
@barbudor
Thank you for help.
I don't think it is redirection, here is the log from curl:
```
curl -v --digest -u "admin:xxxxxxxxx" -g "http://192.168.12.166/cgi-bin/configManager.cgi?action=setConfig&VideoInMode[0].Config[0]=0"
*   Trying 192.168.12.166:80...
* TCP_NODELAY set
* Connected to 192.168.12.166 (192.168.12.166) port 80 (#0)
* Server auth using Digest with user 'admin'
> GET /cgi-bin/configManager.cgi?action=setConfig&VideoInMode[0].Config[0]=0 HTTP/1.1
> Host: 192.168.12.166
> User-Agent: curl/7.68.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 401 Unauthorized
< WWW-Authenticate: Digest realm="Login to b1fb517e19c782ac15a8ee79e124e542", qop="auth", nonce="433829739", opaque="21ba2cc2281212d9bc5762e3f10966b56bf2e164"
< Connection: close
< Set-Cookie:secure; HttpOnly
< CONTENT-LENGTH: 0
< 
* Closing connection 0
* Issue another request to this URL: 'http://192.168.12.166/cgi-bin/configManager.cgi?action=setConfig&VideoInMode[0].Config[0]=0'
* Hostname 192.168.12.166 was found in DNS cache
*   Trying 192.168.12.166:80...
* TCP_NODELAY set
* Connected to 192.168.12.166 (192.168.12.166) port 80 (#1)
* Server auth using Digest with user 'admin'
> GET /cgi-bin/configManager.cgi?action=setConfig&VideoInMode[0].Config[0]=0 HTTP/1.1
> Host: 192.168.12.166
> Authorization: Digest username="admin", realm="Login to b1fb517e19c782ac15a8ee79e124e542", nonce="433829739", uri="/cgi-bin/configManager.cgi?action=setConfig&VideoInMode[0].Config[0]=0", cnonce="YzY4M2ExMjVjMWJjNmNiODM1YmUyY2JlMzdiOTVmODY=", nc=00000001, qop=auth, response="4f92af73dd04ab0bbe3247361b1ddbfe", opaque="21ba2cc2281212d9bc5762e3f10966b56bf2e164"
> User-Agent: curl/7.68.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< X-XSS-Protection: 1;mode=block
< X-Frame-Options: SAMEORIGIN
< Content-Security-Policy: script-src 'self' 'unsafe-inline' 'unsafe-eval'
< Strict-Transport-Security: max-age=604800; includeSubDomains
< Content-type: text/plain;charset=utf-8
< CONNECTION: close
< Set-Cookie:secure; HttpOnly
< CONTENT-LENGTH: 4
< 
OK
* Closing connection 1
```
I tried webquery command ( GET and POST/PUT ) but without success yet - no error log RESULT = {"WebQuery":"Done"}:
```
webquery http://192.168.12.166/cgi-bin/configManager.cgi?action=setConfig&VideoInMode[0].Config[0]=1 GET [Authorization: Basic YWRtaW46eHh4eHh4eA==]
and
webquery http://192.168.12.166/cgi-bin/configManager.cgi?action=setConfig&VideoInMode%5B0%5D.Config%5B0%5D=1 GET [Authorization: Basic YWRtaW46eHh4eHh4eA==]
```
p.s.
If I try curl with basic auth ( encoded or not ) it fail... it ask for digest auth?
```
...
< HTTP/1.1 401 Unauthorized
< WWW-Authenticate: Digest realm="Login to b1fb517e19c782ac15a8ee79e124e542", qop="auth", nonce="869419723", opaque="21ba2cc2281212d9bc5762e3f10966b56bf2e164"
< Connection: close
< Set-Cookie:secure; HttpOnly
< CONTENT-LENGTH: 0
< 
* Closing connection 0
* ```

**Reply by sfromis** [reply]
Since WebQuery supports setting headers, and digest authentication is based on headers with calculated values, it would seem plausible that you could do the same from Tasmota, as long as you do not expect Tasmota to do the cryptographic stuff (MD5)

Example in: https://en.wikipedia.org/wiki/Digest_access_authentication

However, since the response headers from the authentication failure also includes values for digest authentication, I suppose that it would be required to calculate each time, which is out of scope for what feasibly can be done with standard Tasmota.

**Reply by rapi3** [reply]
I am thinking at another approach: to use a tasmota device to send a trigger by mqtt at sunrise/sunset and use raspberry pi subscribed to that topic to do the profile setting with curl...

**Reply by joba-1** [reply]
hm, I’m sure it can be done with tasmota, but if you already have a raspberry pi in the mix, why not doing all there?

**Reply by rapi3** [reply]
I don't know how to calculate the sunrise/sunset on rpi and at some point I will attach a light intensity sensor BH1750 to a wemos running tasmota to do the switching based on exact light.

**Reply by joba-1** [reply]
ok, agreed. Light sensor is the best solution.
Also works for dark rain clouds or solar eclipse and doesn‘t need time or position configuration.

Until then, this code might help :)

http://souptonuts.sourceforge.net/code/sunrise.c.html
